src/workerd/tests/bench-jsstring.c++

```diff
@@ -0,0 +1,302 @@
+      KJ_ASSERT(!concatenated.isFlat());
+      benchmark::DoNotOptimize(concatenated.utf8Length(js));
+    }
+  });
+}
+
+// Benchmarks utf8Length on 8192-char non-flat latin1 strings
+static void JsString_Utf8Length_Latin1_NonFlat_8192(benchmark::State& state) {
+  TestFixture fixture;
+  fixture.runInIoContext([&](const TestFixture::Environment& env) {
+    auto& js = env.js;
+
+    kj::Vector<char> vec1(4096);
+    kj::Vector<char> vec2(4096);
+    for (int i = 0; i < 4096; i++) {
+      vec1.add('a');
+      vec2.add('b');
+    }
+    auto str1 = js.str(vec1.asPtr());
+    auto str2 = js.str(vec2.asPtr());
+
+    for (auto _: state) {
+      auto concatenated = jsg::JsString::concat(js, str1, str2);
+      KJ_ASSERT(!concatenated.isFlat());
+      benchmark::DoNotOptimize(concatenated.utf8Length(js));
+    }
+  });
+}
+
+// Benchmarks utf8Length on 256-char non-flat utf16 strings
+static void JsString_Utf8Length_Utf16_NonFlat_256(benchmark::State& state) {
+  TestFixture fixture;
+  fixture.runInIoContext([&](const TestFixture::Environment& env) {
+    auto& js = env.js;
+
+    kj::Vector<uint16_t> vec1(128);
+    kj::Vector<uint16_t> vec2(128);
+    for (int i = 0; i < 128; i++) {
+      vec1.add(0x1F600 & 0xFFFF);
+      vec2.add(0x1F525 & 0xFFFF);
+    }
+    auto str1 = js.str(vec1.asPtr());
+    auto str2 = js.str(vec2.asPtr());
+
+    for (auto _: state) {
+      auto concatenated = jsg::JsString::concat(js, str1, str2);
+      KJ_ASSERT(!concatenated.isFlat());
+      benchmark::DoNotOptimize(concatenated.utf8Length(js));
+    }
+  });
+}
+
+// Benchmarks utf8Length on 1024-char non-flat utf16 strings
+static void JsString_Utf8Length_Utf16_NonFlat_1024(benchmark::State& state) {
+  TestFixture fixture;
+  fixture.runInIoContext([&](const TestFixture::Environment& env) {
+    auto& js = env.js;
+
+    kj::Vector<uint16_t> vec1(512);
+    kj::Vector<uint16_t> vec2(512);
+    for (int i = 0; i < 512; i++) {
+      vec1.add(0x1F600 & 0xFFFF);
+      vec2.add(0x1F525 & 0xFFFF);
+    }
+    auto str1 = js.str(vec1.asPtr());
+    auto str2 = js.str(vec2.asPtr());
+
+    for (auto _: state) {
+      auto concatenated = jsg::JsString::concat(js, str1, str2);
+      KJ_ASSERT(!concatenated.isFlat());
+      benchmark::DoNotOptimize(concatenated.utf8Length(js));
+    }
+  });
+}
+
+// Benchmarks utf8Length on 8192-char non-flat utf16 strings
+static void JsString_Utf8Length_Utf16_NonFlat_8192(benchmark::State& state) {
+  TestFixture fixture;
+  fixture.runInIoContext([&](const TestFixture::Environment& env) {
+    auto& js = env.js;
+
+    kj::Vector<uint16_t> vec1(4096);
+    kj::Vector<uint16_t> vec2(4096);
+    for (int i = 0; i < 4096; i++) {
+      vec1.add(0x1F600 & 0xFFFF);
+      vec2.add(0x1F525 & 0xFFFF);
+    }
+    auto str1 = js.str(vec1.asPtr());
+    auto str2 = js.str(vec2.asPtr());
+
+    for (auto _: state) {
+      auto concatenated = jsg::JsString::concat(js, str1, str2);
```

Comment:
For SSR, the ConString trees tend to be very deep. The benchmark is only combining two pieces, real world apps will create larger trees that will show a larger performance cliff.